rs-backend/src/server.rs

```diff
@@ -60,309 +142,287 @@ fn get_summary(_req_body: Option<JsonValue>, data: &InputData) -> Result<JsonVal
+        day.by_crate.get(crate_name).map(|krate| {
+            (crate_name, krate)
+        })
+    }).collect::<Vec<_>>();
 
-fn get_data_for_date(day: &TestRun, crates: &JsonValue, phases: &JsonValue, group_by: GroupBy) -> JsonValue {
-    let mut data = JsonValue::new_object();
+    let mut data = HashMap::new();
     if group_by == GroupBy::Crate {
-        for crate_name in crates.members() {
-            let crate_name = crate_name.as_str().unwrap();
-            if let Some(krate) = day.by_crate.get(&*crate_name) {
-                let mut mem = 0;
-                let mut total = 0.0;
-                for phase in phases.members() {
-                    let phase = phase.as_str().unwrap();
-                    if let Some(phase) = krate.get(phase) {
-                        total += phase.time;
-                        mem = max(mem, phase.rss.unwrap());
-                    }
+        for (crate_name, krate) in crates {
+            let mut mem = 0;
+            let mut total = 0.0;
+            for phase in phases {
+                if let Some(phase) = krate.get(phase) {
+                    total += phase.time;
+                    mem = max(mem, phase.rss.unwrap());
                 }
-                data[crate_name] = object! {
-                    "time" => total,
-                    "rss" => mem
-                };
             }
+            data.insert(crate_name, Recording {
+                time: total,
+                rss: mem
+            });
         }
     } else { // group_by == GroupBy::Phase
-        for phase in phases.members() {
-            let phase = phase.as_str().unwrap();
-            let mut total = 0.0;
+        for phase_name in phases {
             let mut mem = 0;
-            for crate_name in crates.members() {
-                let crate_name = crate_name.as_str().unwrap();
-                if let Some(krate) = day.by_crate.get(crate_name) {
-                    if let Some(phase) = krate.get(phase) {
-                        total += phase.time;
-                        mem = max(mem, phase.rss.unwrap());
-                    }
+            let mut total = 0.0;
+            for &(_, krate) in &crates {
+                if let Some(phase) = krate.get(phase_name) {
+                    total += phase.time;
+                    mem = max(mem, phase.rss.unwrap());
                 }
             }
-            data[phase] = object! {
-                "time" => total,
-                "rss" => mem
-            };
+            data.insert(phase_name, Recording {
+                time: total,
+                rss: mem
+            });
         }
     }
 
-    object! {
-        "date" => day.date.format(JS_DATE_FORMAT).to_string(),
-        "commit" => day.commit.clone(),
-        "data" => data
-    }
+    ObjectBuilder::new()
+        .insert("date", day.date.format(JS_DATE_FORMAT).to_string())
+        .insert("commit", day.commit.clone())
+        .insert("data", data)
+        .build()
 }
```

Comment:
I feel like the logic in this function is simplifiable even more, but I'm missing some crucial detail that will get us there.